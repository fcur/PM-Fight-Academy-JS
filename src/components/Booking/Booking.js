import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import { clientsSelector } from "../Clients/Clients.selector";
import {
  formattedTime,
  formattedDuration,
} from "../formattedDate/formattedDate";
import "../Slots/slots.css";
import styles from "../MainStyles/mainStyles.module.css";
import { getClients, resetClient } from "../../redux/Ducks/Clients.duck";
import { getCoaches, resetCoach } from "../../redux/Ducks/Coaches.duck";
import { getServices, resetService } from "../../redux/Ducks/Services.duck";
import { getSlots, resetSlots } from "../../redux/Ducks/Slots.duck";
import {
  deleteBooking,
  getBookings,
  resetBookings,
} from "../../redux/Ducks/Bookings.duck";

const Booking = () => {
  const dispatch = useDispatch();

  // CLIENTS

  const clients = useSelector(clientsSelector);
  const [clientId, setClientId] = useState(1);

  useEffect(() => {
    dispatch(getClients());
    return () => dispatch(resetClient());
  }, [dispatch]);

  const formatClients = () =>
    clients?.map((client) => (
      <option key={client.id} value={client.id}>
        {client.name}
      </option>
    ));

  // COACHES

  const coaches = useSelector((state) => state.coach.data);

  useEffect(() => {
    dispatch(getCoaches());
    return () => dispatch(resetCoach());
  }, [dispatch]);

  const getCoachName = (id) =>
    coaches
      .filter((coach) => coach.id === id)
      .map((coach) => (
        <p key={coach.id}>
          Coach: {coach.firstName} {coach.lastName}
        </p>
      ));

  // SERVICES

  const services = useSelector((state) => state.service.data);
  const [serviceId, setServiceId] = useState(1);

  useEffect(() => {
    dispatch(getServices());
    return () => dispatch(resetService());
  }, [dispatch]);

  const formatServices = () =>
    services?.map((service) => (
      <option key={service.id} value={service.id}>
        {service.name}
      </option>
    ));

  // SLOTS

  const slots = useSelector((state) => state.slots.data);

  useEffect(() => {
    dispatch(getSlots());
    return () => dispatch(resetSlots());
  }, [dispatch]);

  const getCoachFromSlot = (slotId) =>
    slots
      .filter((slot) => slot.id === slotId)
      .map((slot) => <div key={slot.id}>{getCoachName(slot.coachId)}</div>);

  // BOOKING

  const bookings = useSelector((state) => state.booking.data);

  useEffect(() => {
    dispatch(getBookings());
    return () => dispatch(resetBookings());
  }, [dispatch]);

  const deleteHandler = (bookingId) => {
    dispatch(deleteBooking(bookingId));
    dispatch(resetBookings());
  };

  const findBooking = () =>
    bookings
      .filter(
        (booking) =>
          booking.serviceId === serviceId && booking.clientId === clientId
      )
      .map((booking) => {
        return (
          <div key={booking.id} className={styles.bookingBox}>
            <h3>
              {booking.slot.dateStart} - {booking.resultPrice}UAH
            </h3>
            <p>
              {booking.slot.timeStart}-
              {formattedTime(formattedDuration(booking.slot.timeStart))}
            </p>
            {getCoachFromSlot(booking.slot.id)}

            <button
              className={styles.addBtn}
              onClick={() => deleteHandler(booking.id)}
            >
              Delete
            </button>
          </div>
        );
      });

  return (
    <div>
      <div className={styles.bookingSelects}>
        <select onChange={(e) => setServiceId(Number(e.target.value))}>
          {formatServices()}
        </select>
        <select onChange={(e) => setClientId(Number(e.target.value))}>
          {formatClients()}
        </select>
      </div>
      <div className={styles.bookingContainer}>{findBooking()}</div>
    </div>
  );
};

export default Booking;
