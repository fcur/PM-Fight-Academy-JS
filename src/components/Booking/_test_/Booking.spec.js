import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import Booking from "../Booking";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore();

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: () => [
    {
      id: 1,
      slot: {
        id: 2,
        coachId: 3,
        dateStart: "test",
        timeStart: "test",
        duration: "test",
      },
      serviceId: 1,
      clientId: 2,
      resultPrice: 3,
    },
  ],
}));
describe("Booking", () => {
  it("should render properly", () => {
    const { container } = render(
      <Provider store={store}>
        <Booking />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
