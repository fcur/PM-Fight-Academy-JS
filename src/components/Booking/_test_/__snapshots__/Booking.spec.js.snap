// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking should render properly 1`] = `
<div>
  <div>
    <div
      class="bookingSelects"
    >
      <select>
        <option
          value="1"
        />
      </select>
      <select>
        <option
          value="1"
        />
      </select>
    </div>
    <div
      class="bookingContainer"
    />
  </div>
</div>
`;
